File: test/python/test_vineyard.py
# Copyright 2022 Alibaba Group Holding Limited. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

import unittest
import subprocess
import re
import json
import os

from graphlearn_torch.data import *
from graphlearn_torch.distributed import DistDataset

@unittest.skipIf(os.getenv("WITH_VINEYARD", "OFF") == "OFF", "only test with vineyard")
class VineyardDatasetTest(unittest.TestCase):
  sock = "/tmp/vineyard.glt.unittest.sock"

  @classmethod
  def setUpClass(cls):
    cls.vineyardd_process = subprocess.Popen(
      ["vineyardd", "--socket", cls.sock]
    )
    data_dir = os.path.join(
      os.path.dirname(os.path.abspath(__file__)), 'vineyard_data',
      'modern_graph'
    )

    cfg = {
      "vertices":
        [
          {
            "data_path": os.path.join(data_dir, "person.csv"),
            "label": "person",
            "options": "header_row=true&delimiter=|"
          }, {
            "data_path": os.path.join(data_dir, "software.csv"),
            "label": "software",
            "options": "header_row=true&delimiter=|"
          }
        ],
      "edges":
        [
          {
            "data_path": os.path.join(data_dir, "knows.csv"),
            "label": "knows",
            "src_label": "person",
            "dst_label": "person",
            "options": "header_row=true&delimiter=|"
          }, {
            "data_path": os.path.join(data_dir, "created.csv"),
            "label": "created",
            "src_label": "person",
            "dst_label": "software",
            "options": "header_row=true&delimiter=|"
          }
        ],
      "directed": 1,
      "retain_oid": 1,
      "generate_eid": 1,
      "string_oid": 0,
      "local_vertex_map": 0,
      "print_normalized_schema": 1
    }
    json_path = os.path.join(data_dir, "config.json")

    with open(json_path, 'w') as json_file:
      json.dump(cfg, json_file)
    try:
      command = f"vineyard-graph-loader --socket {cls.sock} --config {json_path}"
      output = subprocess.run(
        command, capture_output=True, text=True, shell=True
      )
      match = re.search(r"\[fragment group id\]: (\d+)", str(output))
      if match:
        cls.fid = match.group(1)
      else:
        raise Exception("Fragment Group ID not found.")
    finally:
      os.remove(json_path)

  @classmethod
  def tearDownClass(cls):
    cls.vineyardd_process.kill()

  def setUp(self):
    self.sock = self.__class__.sock
    self.fid = self.__class__.fid
    self.homo_edges = [
      ("person", "knows", "person"),
    ]
    self.homo_edge_weights = {
      ("person", "knows", "person"): "weight",
    }
    self.homo_node_features = {
      "person": ["feat0", "feat1"],
    }
    self.homo_edge_features = {
      ("person", "knows", "person"): ["feat0", "feat1"],
    },
    self.node_labels = {
      "person": "label",
    }

    self.hetero_edges = [
      ("person", "knows", "person"),
      ("person", "created", "software"),
    ]
    self.hetero_edge_weights = {
      ("person", "knows", "person"): "weight",
      ("person", "created", "software"): "weight",
    }
    self.hetero_node_features = {
      "person": ["feat0", "feat1"],
      "software": ["feat0"],
    }
    self.hetero_edge_features = {
      ("person", "knows", "person"): ["feat0", "feat1"],
      ("person", "created", "software"): ["feat0"],
    }

  def test_homo_dataset(self):
    ds = Dataset()
    ds.load_vineyard(
      self.fid,
      self.sock,
      edges=self.homo_edges,
      edge_weights=self.homo_edge_weights,
      node_features=self.homo_node_features,
      edge_features=self.homo_edge_features,
      node_labels=self.node_labels,
    )
    self.assertEqual(ds.graph.row_count, 4)
    self.assertEqual(ds.graph.col_count, 2)
    self.assertEqual(ds.graph.edge_count, 2)
    self.assertEqual(ds.graph.topo.edge_weights.shape, (2,))
    self.assertEqual(ds.node_features.shape, (4, 2))
    self.assertEqual(ds.edge_features.shape, (2, 2))
    self.assertEqual(ds.node_labels.shape, (4,))

  def test_in_homo_dataset(self):
    ds = Dataset(edge_dir="in")
    ds.load_vineyard(
      self.fid,
      self.sock,
      edges=self.homo_edges,
    )
    self.assertEqual(ds.graph.row_count, 4)
    self.assertEqual(ds.graph.col_count, 1)
    self.assertEqual(ds.graph.edge_count, 2)

  def test_hetero_dataset(self):
    ds = Dataset()
    ds.load_vineyard(
      self.fid,
      self.sock,
      edges=self.hetero_edges,
      edge_weights=self.hetero_edge_weights,
      node_features=self.hetero_node_features,
      edge_features=self.hetero_edge_features,
      node_labels=self.node_labels,
    )
    graph1 = ds.graph[("person", "knows", "person")]
    graph2 = ds.graph[("person", "created", "software")]
    self.assertEqual(graph1.row_count, 4)
    self.assertEqual(graph1.col_count, 2)
    self.assertEqual(graph1.edge_count, 2)
    self.assertEqual(graph1.topo.edge_weights.shape, (2,))

    self.assertEqual(graph2.row_count, 4)
    self.assertEqual(graph2.col_count, 2)
    self.assertEqual(graph2.edge_count, 4)
    self.assertEqual(graph2.topo.edge_weights.shape, (4,))

    self.assertEqual(ds.node_features["person"].shape, (4, 2))
    self.assertEqual(ds.node_features["software"].shape, (2, 1))
    self.assertEqual(
      ds.edge_features[("person", "knows", "person")].shape, (2, 2)
    )
    self.assertEqual(
      ds.edge_features[("person", "created", "software")].shape, (4, 1)
    )
    self.assertEqual(ds.node_labels["person"].shape, (4,))

  def test_in_hetero_dataset(self):
    ds = Dataset(edge_dir="in")
    ds.load_vineyard(
      self.fid,
      self.sock,
      edges=self.hetero_edges,
    )
    graph1 = ds.graph[("person", "knows", "person")]
    graph2 = ds.graph[("person", "created", "software")]

    self.assertEqual(graph1.row_count, 4)
    self.assertEqual(graph1.col_count, 1)
    self.assertEqual(graph1.edge_count, 2)

    self.assertEqual(graph2.row_count, 2)
    self.assertEqual(graph2.col_count, 3)
    self.assertEqual(graph2.edge_count, 4)

  def test_homo_dist_dataset(self):
    ds = DistDataset()
    ds.load_vineyard(
      self.fid,
      self.sock,
      edges=self.homo_edges,
      node_features=self.homo_node_features,
      edge_features=self.homo_edge_features,
      node_labels=self.node_labels,
    )
    self.assertEqual(ds.node_pb.shape, (4,))
    self.assertEqual(ds.edge_pb.shape, (2,))
    self.assertEqual(ds.node_feat_pb.shape, (4,))
    self.assertEqual(ds.edge_feat_pb.shape, (2,))

  def test_hetero_dist_dataset(self):
    ds = DistDataset()
    ds.load_vineyard(
      self.fid,
      self.sock,
      edges=self.hetero_edges,
      node_features=self.hetero_node_features,
      edge_features=self.hetero_edge_features,
      node_labels=self.node_labels,
    )
    self.assertEqual(ds.node_pb["person"].shape, (4,))
    # self.assertEqual(ds.node_pb["software"].shape, (2,))

    self.assertEqual(ds.edge_pb[("person", "knows", "person")].shape, (2,))
    self.assertEqual(ds.edge_pb[("person", "created", "software")].shape, (4,))

    self.assertEqual(ds.node_feat_pb["person"].shape, (4,))
    self.assertEqual(ds.node_feat_pb["software"].shape, (2,))

    self.assertEqual(ds.edge_feat_pb[("person", "knows", "person")].shape, (2,))
    self.assertEqual(
      ds.edge_feat_pb[("person", "created", "software")].shape, (4,)
    )


if __name__ == "__main__":
  unittest.main()

File: test/cpp/test_vineyard.cc
/* Copyright 2022 Alibaba Group Holding Limited. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "gtest/gtest.h"

#include "graphlearn_torch/v6d/vineyard_utils.h"

#include <grape/communication/communicator.h>
#include <grape/worker/comm_spec.h>

using namespace graphlearn_torch::vineyard_utils;

class VineyardTest : public ::testing::Test {
protected:
  void SetUp() override {
    ipc_socket = "/var/run/vineyard.sock";
    object_id_str = "29791161733259207";
    v_label_name = "person";
    e_label_name = "knows";
    vcols_1 = {"feat0"};
    vcols_2 = {"feat0", "feat1"};
    ecols_1 = {"feat0"};
    ecols_2 = {"feat0", "feat1"};
  }

protected:
  std::string ipc_socket;
  std::string object_id_str;
  std::string v_label_name;
  std::string e_label_name;
  std::vector<std::string> vcols_1;
  std::vector<std::string> vcols_2;
  std::vector<std::string> ecols_1;
  std::vector<std::string> ecols_2;
};

TEST_F(VineyardTest, VineyardTest) {

  grape::InitMPIComm();
  {
    grape::CommSpec comm_spec;
    comm_spec.Init(MPI_COMM_WORLD);
    torch::Tensor out_indptr;
    torch::Tensor out_indices;
    torch::Tensor out_edge_ids;
    std::tie(out_indptr, out_indices, out_edge_ids) = ToCSR(ipc_socket, object_id_str, v_label_name, e_label_name, "out", true);
    // std::cout << "ToCSR out" << std::endl;
    // std::cout << "out_indptr" <<  out_indptr << std::endl;
    // std::cout << "out_indices" <<  out_indices << std::endl;
    // std::cout << "out_edge_ids" <<  out_edge_ids << std::endl;
    
    torch::Tensor in_indptr;
    torch::Tensor in_indices;
    torch::Tensor in_edge_ids;
    std::tie(in_indptr, in_indices, in_edge_ids) = ToCSR(ipc_socket, object_id_str, v_label_name, e_label_name, "in", true);
    // std::cout << "ToCSR in" << std::endl;
    // std::cout << "in_indptr" <<  in_indptr << std::endl;
    // std::cout << "in_indices" <<  in_indices << std::endl;
    // std::cout << "in_edge_ids" <<  in_edge_ids << std::endl;

    ASSERT_EQ(out_indptr.numel(), 5);
    ASSERT_EQ(out_indptr.dtype(), torch::kLong);
    ASSERT_TRUE(torch::allclose(out_indptr, torch::tensor({0, 0, 0, 0, 2}, torch::kLong)));
    
    ASSERT_EQ(out_indices.numel(), 2);
    ASSERT_EQ(out_indices.dtype(), torch::kLong);
    ASSERT_TRUE(torch::allclose(out_indices, torch::tensor({0, 2}, torch::kLong)));
    
    ASSERT_EQ(out_edge_ids.numel(), 2);
    ASSERT_EQ(out_edge_ids.dtype(), torch::kLong);
    ASSERT_TRUE(torch::allclose(out_edge_ids, torch::tensor({0, 1}, torch::kLong)));
    
    ASSERT_EQ(in_indptr.numel(), 5);
    ASSERT_EQ(in_indptr.dtype(), torch::kLong);
    ASSERT_TRUE(torch::allclose(in_indptr, torch::tensor({0, 1, 1, 2, 2}, torch::kLong)));
    
    ASSERT_EQ(in_indices.numel(), 2);
    ASSERT_EQ(in_indices.dtype(), torch::kLong);
    ASSERT_TRUE(torch::allclose(in_indices, torch::tensor({3, 3}, torch::kLong)));
    
    ASSERT_EQ(in_edge_ids.numel(), 2);
    ASSERT_EQ(in_edge_ids.dtype(), torch::kLong);
    ASSERT_TRUE(torch::allclose(in_edge_ids, torch::tensor({0, 1}, torch::kLong)));


    auto vfeat_1 = LoadVertexFeatures(ipc_socket, object_id_str, v_label_name, vcols_1);
    ASSERT_EQ(vfeat_1.sizes(), torch::IntArrayRef({4, 1}));
    ASSERT_EQ(vfeat_1.dtype(), torch::kDouble);

    auto efeat_1 = LoadEdgeFeatures(ipc_socket, object_id_str, e_label_name, ecols_1);
    ASSERT_EQ(efeat_1.sizes(), torch::IntArrayRef({2, 1}));
    ASSERT_EQ(efeat_1.dtype(), torch::kDouble);
    
    auto vfeat_2 = LoadVertexFeatures(ipc_socket, object_id_str, v_label_name, vcols_2);
    ASSERT_EQ(vfeat_2.sizes(), torch::IntArrayRef({4, 2}));
    ASSERT_EQ(vfeat_2.dtype(), torch::kDouble);

    auto efeat_2 = LoadEdgeFeatures(ipc_socket, object_id_str, e_label_name, ecols_2);
    ASSERT_EQ(efeat_2.sizes(), torch::IntArrayRef({2, 2}));
    ASSERT_EQ(efeat_2.dtype(), torch::kDouble);
  }
}